runners/direct-java/src/main/java/org/apache/beam/runners/direct/DirectRunner.java

```diff
@@ -76,19 +77,38 @@
    * type of transform it is overriding.
    */
   @SuppressWarnings("rawtypes")
-  private static Map<Class<? extends PTransform>, PTransformOverrideFactory>
-      defaultTransformOverrides =
-          ImmutableMap.<Class<? extends PTransform>, PTransformOverrideFactory>builder()
-              .put(CreatePCollectionView.class, new ViewOverrideFactory())
-              .put(GroupByKey.class, new DirectGroupByKeyOverrideFactory())
-              .put(TestStream.class, new DirectTestStreamFactory())
-              .put(Write.Bound.class, new WriteWithShardingFactory())
-              .put(ParDo.Bound.class, new ParDoSingleViaMultiOverrideFactory())
-              .put(ParDo.BoundMulti.class, new ParDoMultiOverrideFactory())
-              .put(
-                  SplittableParDo.GBKIntoKeyedWorkItems.class,
-                  new DirectGBKIntoKeyedWorkItemsOverrideFactory())
-              .build();
+  private static Map<PTransformMatcher, PTransformOverrideFactory> defaultTransformOverrides =
+      ImmutableMap.<PTransformMatcher, PTransformOverrideFactory>builder()
+          .put(
+              PTransformMatchers.classEqualTo(Bound.class),
+              new WriteWithShardingFactory()) /* Uses a view internally. */
+          .put(
+              PTransformMatchers.classEqualTo(CreatePCollectionView.class),
+              new ViewOverrideFactory()) /* Uses pardos and GBKs */
+          .put(
+              PTransformMatchers.classEqualTo(TestStream.class),
+              new DirectTestStreamFactory()) /* primitive */
+          // SplittableParMultiDo is implemented
+          .put(PTransformMatchers.splittableParDoSingle(), new ParDoSingleViaMultiOverrideFactory())
+          .put(
+              PTransformMatchers.stateOrTimerParDoSingle(),
+              new ParDoSingleViaMultiOverrideFactory())
+          .put(PTransformMatchers.splittableParDoMulti(), new ParDoMultiOverrideFactory())
```

Comment:
When you have a sensitivity to ordering, comment about it it. For example "ParDoSingleViaMulti comes before ParDoMultiOverrideFactory because state and splittable are only set up to work on multi" and then later "ParDoSingleViaMulti comes _after_ ParDoMultiOverrideFactory because the resulting expansions contains yet more ParDo singles"